File: lib/GeoCalCore/Implementation/hdf_orbit_test.cc
#include "unit_test_support.h"
#include "hdf_orbit.h"
#include "eci_tod.h"
#include "quaternion_camera.h"
#include "geodetic.h"
using namespace GeoCal;
using namespace blitz;

class HdfOrbitFixture : public GlobalFixture {
public:
  HdfOrbitFixture() 
  {
    std::string fname = test_data_dir() + "sample_orbit.h5";
    orb.reset(new HdfOrbit<EciTod, TimeAcsCreator>(fname));
    t = Time::time_acs(215077459.472);
  }
  boost::shared_ptr<HdfOrbit<EciTod, TimeAcsCreator> > orb;
  Time t;
};

BOOST_FIXTURE_TEST_SUITE(hdf_orbit, HdfOrbitFixture)

// This attempts to match the results of the sc2rpc unit tests.
// **NOTE** We get results that are about 7 meters different than what
// sc2rpc gets. This was tracked down to what I believe is a more
// accurate calculation the spice toolkit for J2000 to ECEF.  We allow
// a 10 meter error in comparing to sc2rpc.

BOOST_AUTO_TEST_CASE(sc2rpc)
{
  QuaternionCamera cam(boost::math::quaternion<double>(1,0,0,0),
		       3375, 3648,
		       1.0 / 2500000,
		       1.0 / 2500000,
		       1.0,
		       FrameCoordinate(1688.0, 1824.5),
		       QuaternionCamera::LINE_IS_Y);
  boost::shared_ptr<CartesianFixed> pt =
    orb->reference_surface_intersect_approximate(t, cam, FrameCoordinate(3375, 3648));
  Geodetic pt_geod(-60.3162137, 47.2465154);
  BOOST_CHECK(distance(pt_geod, *pt) < 10.0);
}

BOOST_AUTO_TEST_CASE(basic)
{
  Geodetic gexpect(-59.6722, 45.2541, 661831);
  BOOST_CHECK(distance(*orb->position_cf(t), gexpect) < 10);
  boost::array<double, 3> p1 = orb->position_cf(t)->position;
  double tdelta = 1.0;
  boost::array<double, 3> p2 = orb->position_cf(t + tdelta)->position;
  boost::shared_ptr<QuaternionOrbitData> od = 
    boost::dynamic_pointer_cast<QuaternionOrbitData>(orb->orbit_data(t));
  boost::array<double, 3> v = od->velocity_cf();
  for(int i = 0; i < 3; ++i)
    BOOST_CHECK(fabs((p2[i] - p1[i]) / tdelta - v[i]) < 1.0);
}

BOOST_AUTO_TEST_CASE(derivative_ci)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  boost::shared_ptr<CartesianInertial> p0 = orb->position_ci(t2.value());
  boost::shared_ptr<CartesianInertial> p1 = orb->position_ci(t2.value() + eps);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1->position[0] - p0->position[0]) / eps;
  jac_fd(1,0) = (p1->position[1] - p0->position[1]) / eps;
  jac_fd(2,0) = (p1->position[2] - p0->position[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  boost::array<AutoDerivative<double>, 3> p2 = 
    orb->position_ci_with_derivative(t2);
  jac_calc(0,Range::all()) = p2[0].gradient();
  jac_calc(1,Range::all()) = p2[1].gradient();
  jac_calc(2,Range::all()) = p2[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}

BOOST_AUTO_TEST_CASE(derivative_cf)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  boost::shared_ptr<CartesianFixed> p0 = orb->position_cf(t2.value());
  boost::shared_ptr<CartesianFixed> p1 = orb->position_cf(t2.value() + eps);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1->position[0] - p0->position[0]) / eps;
  jac_fd(1,0) = (p1->position[1] - p0->position[1]) / eps;
  jac_fd(2,0) = (p1->position[2] - p0->position[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  boost::array<AutoDerivative<double>, 3> p2 = 
    orb->position_cf_with_derivative(t2);
  jac_calc(0,Range::all()) = p2[0].gradient();
  jac_calc(1,Range::all()) = p2[1].gradient();
  jac_calc(2,Range::all()) = p2[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}

BOOST_AUTO_TEST_CASE(derivative_ci_look)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  SimpleCamera cam;
  ScLookVector sl = cam.sc_look_vector(FrameCoordinate(0,0), 0);
  ScLookVectorWithDerivative sl_wd = 
    cam.sc_look_vector_with_derivative(FrameCoordinateWithDerivative(0,0), 0);
  CartesianInertialLookVector p0 = orb->ci_look_vector(t2.value(), sl);
  CartesianInertialLookVector p1 = orb->ci_look_vector(t2.value() + eps, sl);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1.look_vector[0] - p0.look_vector[0]) / eps;
  jac_fd(1,0) = (p1.look_vector[1] - p0.look_vector[1]) / eps;
  jac_fd(2,0) = (p1.look_vector[2] - p0.look_vector[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  CartesianInertialLookVectorWithDerivative p2 = 
    orb->ci_look_vector(t2, sl_wd);
  jac_calc(0,Range::all()) = p2.look_vector[0].gradient();
  jac_calc(1,Range::all()) = p2.look_vector[1].gradient();
  jac_calc(2,Range::all()) = p2.look_vector[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}

BOOST_AUTO_TEST_CASE(derivative_cf_look)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  SimpleCamera cam;
  ScLookVector sl = cam.sc_look_vector(FrameCoordinate(0,0), 0);
  ScLookVectorWithDerivative sl_wd = 
    cam.sc_look_vector_with_derivative(FrameCoordinateWithDerivative(0,0), 0);
  CartesianFixedLookVector p0 = orb->cf_look_vector(t2.value(), sl);
  CartesianFixedLookVector p1 = orb->cf_look_vector(t2.value() + eps, sl);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1.look_vector[0] - p0.look_vector[0]) / eps;
  jac_fd(1,0) = (p1.look_vector[1] - p0.look_vector[1]) / eps;
  jac_fd(2,0) = (p1.look_vector[2] - p0.look_vector[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  CartesianFixedLookVectorWithDerivative p2 = 
    orb->cf_look_vector(t2, sl_wd);
  jac_calc(0,Range::all()) = p2.look_vector[0].gradient();
  jac_calc(1,Range::all()) = p2.look_vector[1].gradient();
  jac_calc(2,Range::all()) = p2.look_vector[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}

BOOST_AUTO_TEST_CASE(derivative_sc_look1)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  SimpleCamera cam;
  ScLookVector sl = cam.sc_look_vector(FrameCoordinate(0,0), 0);
  CartesianInertialLookVector clv = orb->ci_look_vector(t2.value(), sl);
  CartesianInertialLookVectorWithDerivative 
    clv_wd(clv.look_vector[0],clv.look_vector[1],clv.look_vector[2]);
  ScLookVector p0 = orb->sc_look_vector(t2.value(), clv);
  ScLookVector p1 = orb->sc_look_vector(t2.value() + eps, clv);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1.look_vector[0] - p0.look_vector[0]) / eps;
  jac_fd(1,0) = (p1.look_vector[1] - p0.look_vector[1]) / eps;
  jac_fd(2,0) = (p1.look_vector[2] - p0.look_vector[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  ScLookVectorWithDerivative p2 = 
    orb->sc_look_vector(t2, clv_wd);
  jac_calc(0,Range::all()) = p2.look_vector[0].gradient();
  jac_calc(1,Range::all()) = p2.look_vector[1].gradient();
  jac_calc(2,Range::all()) = p2.look_vector[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}

BOOST_AUTO_TEST_CASE(derivative_sc_look2)
{
  TimeWithDerivative t2 = 
    TimeWithDerivative::time_pgs(AutoDerivative<double>(t.pgs(), 0, 1));
  double eps = 1e-3;
  SimpleCamera cam;
  ScLookVector sl = cam.sc_look_vector(FrameCoordinate(0,0), 0);
  CartesianFixedLookVector clv = orb->cf_look_vector(t2.value(), sl);
  CartesianFixedLookVectorWithDerivative 
    clv_wd(clv.look_vector[0],clv.look_vector[1],clv.look_vector[2]);
  ScLookVector p0 = orb->sc_look_vector(t2.value(), clv);
  ScLookVector p1 = orb->sc_look_vector(t2.value() + eps, clv);
  Array<double, 2> jac_fd(3,1);
  jac_fd(0,0) = (p1.look_vector[0] - p0.look_vector[0]) / eps;
  jac_fd(1,0) = (p1.look_vector[1] - p0.look_vector[1]) / eps;
  jac_fd(2,0) = (p1.look_vector[2] - p0.look_vector[2]) / eps;
  Array<double, 2> jac_calc(3, 1);
  ScLookVectorWithDerivative p2 = 
    orb->sc_look_vector(t2, clv_wd);
  jac_calc(0,Range::all()) = p2.look_vector[0].gradient();
  jac_calc(1,Range::all()) = p2.look_vector[1].gradient();
  jac_calc(2,Range::all()) = p2.look_vector[2].gradient();
  BOOST_CHECK_MATRIX_CLOSE_TOL(jac_fd, jac_calc, 0.1);
}


BOOST_AUTO_TEST_CASE(serialization)
{
  if(!have_serialize_supported())
    return;
  std::string d = serialize_write_string(orb);
  if(false)
    std::cerr << d;
  boost::shared_ptr<Orbit> orbr = 
    serialize_read_string<Orbit>(d);
  BOOST_CHECK(fabs(orb->min_time() - orbr->min_time()) < 1e-3);
  BOOST_CHECK(fabs(orb->max_time() - orbr->max_time()) < 1e-3);
}

BOOST_AUTO_TEST_SUITE_END()

File: python/lib/hdf_orbit_test.py
from __future__ import division
from future import standard_library
standard_library.install_aliases()
from builtins import range
from past.utils import old_div
import h5py
from test_support import *
import pickle

def create_sample_file():
    '''This create a HDF 5 file that contains orbit information. We can use
    this is a base for converting from some other format to HDF 5 (e.g., 
    reading Level 0).'''
    # Create file, overwriting any existing file
    f = h5py.File(unit_test_data + "sample_orbit.h5", "w")
    # Create group, so we could have other things in the HDF 5 if we wanted
    # without worrying about possible name classes
    g = f.create_group("Orbit")
    # Separate ephemeris and attitude
    eph = g.create_group("Ephemeris")
    att = g.create_group("Attitude")
    # Easiest if we know ahead of time what size we want to write. We
    # can have resizable arrays if needed.
    tm = eph.create_dataset("Time", (100,), "=f8")
    pos = eph.create_dataset("Position", (100,3), "=f8")
    vel = eph.create_dataset("Velocity", (100,3), "=f8")
    # Include descriptions
    tm.attrs["Description"] = "This is ACS time that position is measured"
    pos.attrs["Description"] = "This is ECI True of Date Position, in meters"
    vel.attrs["Description"] = "This is the velocity that goes with the position, in meter/second"
    # Repeat with attitude
    tm2 = att.create_dataset("Time", (1000,), "=f8")
    quat = att.create_dataset("Quaternion", (1000,4), "=f8")
    tm2.attrs["Description"] = "This is ACS time that attitude quaternion is measured"
    quat.attrs["Description"] = "This is the attitude quaternion. This gives the orientation of the spacecraft with the ECI TOD position"

    # This isn't the best test data, but stick the spacecraft at one
    # unmoving position, and one orientation. This matches the data
    # used in the quaternion_camera_test.cc, which is turn matches
    # the unit test of sc2rpc. We can come up with better data at some
    # point if needed.
    # Time of interest is 215077459.472
    for i, t in enumerate(range(215077450 - 50 * 60, 215077450 + 50 * 60, 60)):
        tm[i] = t
        pos[i,:] = np.array([3435100.496, 945571.538, -6053387.573])
        vel[i,:] = np.array([0.0,0.0,0.0])
    for i, t in enumerate(range(215077450 - 500 * 1, 215077450 + 500 * 1, 1)):
        tm2[i] = t
        quat[i, :] = [0.946366, 0.0, -0.323096813, 0.0]

# create_sample_file()

@require_hdf5
def test_sc2rpc():
    '''This attempts to match the results of the sc2rpc unit tests.
    **NOTE** We get results that are about 7 meters different than what
    sc2rpc gets. This was tracked down to what I believe is a more
    accurate calculation the spice toolkit for J2000 to ECEF.  We allow
    a 10 meter error in comparing to sc2rpc.'''
    orb = HdfOrbit_EciTod_TimeAcs(unit_test_data + "sample_orbit.h5")
    cam = QuaternionCamera(Quaternion_double(1,0,0,0),
                           3375, 3648,
                           old_div(1.0, 2500000),
                           old_div(1.0, 2500000),
                           1.0,
                           FrameCoordinate(1688.0, 1824.5),
                           QuaternionCamera.LINE_IS_Y)
    t = Time.time_acs(215077459.472);
    pt = orb.reference_surface_intersect_approximate(t, cam,
                                            FrameCoordinate(3375, 3648))
    pt_geod = Geodetic(-60.3162137, 47.2465154)
    assert distance(pt_geod, pt) < 10.0
    

@require_serialize
@require_hdf5
def test_pickle():
    orb = HdfOrbit_EciTod_TimeAcs(unit_test_data + "sample_orbit.h5")
    t = pickle.dumps(orb, pickle.HIGHEST_PROTOCOL)
    orb2 = pickle.loads(t)
    cam = QuaternionCamera(Quaternion_double(1,0,0,0),
                           3375, 3648,
                           old_div(1.0, 2500000),
                           old_div(1.0, 2500000),
                           1.0,
                           FrameCoordinate(1688.0, 1824.5),
                           QuaternionCamera.LINE_IS_Y)
    t = Time.time_acs(215077459.472);
    pt = orb.reference_surface_intersect_approximate(t, cam, FrameCoordinate(3375, 3648))
    pt2 = orb2.reference_surface_intersect_approximate(t, cam, FrameCoordinate(3375, 3648))
    assert distance(pt, pt2) < 0.01
    od1 = orb.orbit_data(t)
    pick = pickle.dumps(od1, pickle.HIGHEST_PROTOCOL)
    od2 = pickle.loads(pick)
    pt = od1.reference_surface_intersect_approximate(cam, FrameCoordinate(3375, 3648))
    pt2 = od2.reference_surface_intersect_approximate(cam, FrameCoordinate(3375, 3648))
    assert distance(pt, pt2) < 0.01
    
    
